contracts/xasset/src/token.rs

```diff
@@ -783,129 +1493,118 @@ impl IsCollateralized for Token {
-        Token::init(
-            xlm_sac,
-            xlm_contract,
-            asset_contract,
-            pegged_asset,
-            min_collat_ratio,
-            name,
-            symbol,
-            decimals, // FIXME: we should use this instead of the data feed to get xasset decimals?
-            annual_interest_rate,
-        );
-    }
-    fn set_xlm_sac(&mut self, to: Address) {
-        self::Contract::require_auth();
-        self.xlm_sac.set(&to);
-    }
-    fn set_xlm_contract(&mut self, to: Address) {
-        self::Contract::require_auth();
-        self.xlm_contract.set(&to);
-    }
-    fn set_asset_contract(&mut self, to: Address) {
-        self::Contract::require_auth();
-        self.asset_contract.set(&to);
-    }
-    fn set_pegged_asset(&mut self, to: Symbol) {
-        self::Contract::require_auth();
-        self.pegged_asset.set(&to);
-    }
-    fn set_min_collat_ratio(&mut self, to: u32) -> u32 {
-        self::Contract::require_auth();
-        self.min_collat_ratio.set(&to);
+#[contractimpl]
+impl IsCDPAdmin for TokenContract {
+    fn set_xlm_sac(env: &Env, to: Address) {
+        Self::require_admin(env);
+        let mut state = TokenStorage::get_state(env);
+        state.xlm_sac = to;
+        TokenStorage::set_state(env, &state);
+    }
+    fn set_xlm_contract(env: &Env, to: Address) {
+        Self::require_admin(env);
+        let mut state = TokenStorage::get_state(env);
+        state.xlm_sac = to;
+        TokenStorage::set_state(env, &state);
+    }
+    fn set_asset_contract(env: &Env, to: Address) {
+        Self::require_admin(env);
+        let mut state = TokenStorage::get_state(env);
+        state.asset_contract = to;
+        TokenStorage::set_state(env, &state);
+    }
+    fn set_pegged_asset(env: &Env, to: Symbol) {
+        Self::require_admin(env);
+        let mut state = TokenStorage::get_state(env);
+        state.pegged_asset = to;
+        TokenStorage::set_state(env, &state);
+    }
+    fn set_min_collat_ratio(env: &Env, to: u32) -> u32 {
+        Self::require_admin(env);
+        let mut state = TokenStorage::get_state(env);
+        state.min_collat_ratio = to;
+        TokenStorage::set_state(env, &state);
         to
     }
-    fn set_interest_rate(&mut self, new_rate: u32) -> u32 {
-        self::Contract::require_auth();
-        self.set_annual_interest_rate(new_rate);
+
+    fn set_interest_rate(env: &Env, new_rate: u32) -> u32 {
+        Self::require_admin(env);
+        Self::set_annual_interest_rate(env, new_rate);
         new_rate
     }
 
-    fn get_interest_rate(&self) -> u32 {
-        self.get_annual_interest_rate()
+    fn get_interest_rate(env: &Env) -> u32 {
+        Self::get_annual_interest_rate(env)
     }
 
-    fn get_total_interest_collected(&self) -> i128 {
-        self.get_total_interest_collected()
+    fn get_total_interest_collected(env: &Env) -> i128 {
+        TokenStorage::get_state(env).interest_collected.clone()
```

Comment:
`i128` and all number types implement `Copy` meaning that their bytes are copied automatically.
```suggestion
        TokenStorage::get_state(env).interest_collected
```